File: src/Arma3ServerTools.Application/Services/BansService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;using Arma3ServerTools.Core;
using Arma3ServerTools.Core.Models;

namespace Arma3ServerTools.Application.Services
{
    public sealed class BansService
    {
        public static readonly Encoding Utf8NoBom = new UTF8Encoding(false);

        public IReadOnlyList<LocalBansEntity> LoadLocalBans(string serverDir, string serverUuid)
        {
            var result = new List<LocalBansEntity>();
            var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            foreach (string path in GetBanFilePaths(serverDir, serverUuid))
            {
                if (!File.Exists(path))
                {
                    continue;
                }

                string content = File.ReadAllText(path, Utf8NoBom);
                AppendParsedBans(content, result, seen, true, string.Empty);
            }

            return result;
        }

        public IReadOnlyList<LocalBansEntity> FetchRemoteBans(string url)
        {
            var result = new List<LocalBansEntity>();
            if (string.IsNullOrWhiteSpace(url))
            {
                return result;
            }

            using (var client = new WebClient())
            {
                client.Encoding = Utf8NoBom;
                string content = client.DownloadString(url);
                AppendParsedBans(content, result, new HashSet<string>(StringComparer.OrdinalIgnoreCase), false, url);
            }

            return result;
        }

        public IReadOnlyList<LocalBansEntity> FetchRemoteBansFromUrls(IEnumerable<BansUrlEntity> urls)
        {
            var result = new List<LocalBansEntity>();
            var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            if (urls == null)
            {
                return result;
            }

            foreach (BansUrlEntity entry in urls)
            {
                if (entry == null || !entry.enable || string.IsNullOrWhiteSpace(entry.url))
                {
                    continue;
                }

                try
                {
                    using (var client = new WebClient())
                    {
                        client.Encoding = Utf8NoBom;
                        string content = client.DownloadString(entry.url);
                        AppendParsedBans(content, result, seen, false, entry.url);
                    }
                }
                catch
                {
                    // 单个 URL 失败不影响其余列表。
                }
            }

            return result;
        }

        public OperationResult SaveLocalBans(string serverDir, string serverUuid, IEnumerable<LocalBansEntity> bans)
        {
            var builder = new StringBuilder();
            foreach (LocalBansEntity ban in bans)
            {
                if (ban == null || string.IsNullOrWhiteSpace(ban.GUID))
                {
                    continue;
                }

                string expiry = ban.Time;
                if (string.Equals(expiry, "永久封禁", StringComparison.OrdinalIgnoreCase))
                {
                    expiry = "-1";
                }

                builder.Append(ban.GUID).Append(' ').Append(expiry).Append(' ').AppendLine(ban.Reason ?? string.Empty);
            }

            string payload = builder.ToString();
            foreach (string path in GetBanFilePaths(serverDir, serverUuid))
            {
                try
                {
                    string directory = Path.GetDirectoryName(path);
                    if (!string.IsNullOrEmpty(directory))
                    {
                        Directory.CreateDirectory(directory);
                    }

                    File.WriteAllText(path, payload, Utf8NoBom);
                }
                catch (Exception ex)
                {
                    return OperationResult.Fail("保存封禁列表失败 [" + path + "]: " + ex.Message);
                }
            }

            return OperationResult.Ok();
        }

        private static IEnumerable<string> GetBanFilePaths(string serverDir, string serverUuid)
        {
            yield return serverDir + @"\bans.txt";
            yield return serverDir + @"\destiny_serverconfig\" + serverUuid + @"\BattlEye\bans.txt";
            yield return serverDir + @"\destiny_serverconfig\" + serverUuid + @"\Users\" + serverUuid + @"\bans.txt";
        }

        private static void AppendParsedBans(
            string data,
            List<LocalBansEntity> target,
            HashSet<string> seen,
            bool local,
            string sourceName)
        {
            if (string.IsNullOrEmpty(data))
            {
                return;
            }

            string[] lines = data.Split(new[] { "\r\n", "\n" }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string line in lines)
            {
                string[] parts = line.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 0 || string.IsNullOrWhiteSpace(parts[0]))
                {
                    continue;
                }

                if (!seen.Add(parts[0]))
                {
                    continue;
                }

                string expiry = parts.Length > 1 ? parts[1] : string.Empty;
                if (string.Equals(expiry, "-1", StringComparison.Ordinal))
                {
                    expiry = "永久封禁";
                }

                string reason = parts.Length > 2 ? parts[2] : string.Empty;
                if (local)
                {
                    target.Add(new LocalBansEntity(parts[0], expiry, reason, string.Empty, string.Empty));
                }
                else
                {
                    string addTime = parts.Length > 3 ? parts[3] : string.Empty;
                    string syncName = parts.Length > 4 ? parts[4] : sourceName;
                    target.Add(new LocalBansEntity(parts[0], expiry, reason, addTime, syncName));
                }
            }
        }
    }
}

File: packages/service/src/bans/bans-service.ts
import * as fs from "node:fs";
import * as path from "node:path";

export interface LocalBanEntry {
  guid: string;
  time: string;
  reason: string;
  name?: string;
  ip?: string;
}

const SERVER_CONFIG_FOLDER = "serverConfig";

export function getBanFilePaths(serverDir: string, serverUuid: string): string[] {
  return [
    path.join(serverDir, "bans.txt"),
    path.join(serverDir, SERVER_CONFIG_FOLDER, serverUuid, "BattlEye", "bans.txt"),
    path.join(serverDir, SERVER_CONFIG_FOLDER, serverUuid, "Users", serverUuid, "bans.txt"),
  ];
}

export function loadLocalBans(serverDir: string, serverUuid: string): LocalBanEntry[] {
  const result: LocalBanEntry[] = [];
  const seen = new Set<string>();

  for (const filePath of getBanFilePaths(serverDir, serverUuid)) {
    if (!fs.existsSync(filePath)) {
      continue;
    }
    try {
      const content = fs.readFileSync(filePath, "utf-8");
      appendParsedBans(content, result, seen);
    } catch {
      // skip unreadable
    }
  }

  return result;
}

export function saveLocalBans(
  serverDir: string,
  serverUuid: string,
  bans: LocalBanEntry[]
): { success: boolean; message: string } {
  const lines: string[] = [];
  for (const ban of bans) {
    if (!ban.guid?.trim()) {
      continue;
    }
    let expiry = ban.time ?? "-1";
    if (expiry === "永久封禁") {
      expiry = "-1";
    }
    lines.push(`${ban.guid.trim()} ${expiry} ${ban.reason ?? ""}`.trimEnd());
  }
  const payload = lines.length > 0 ? `${lines.join("\r\n")}\r\n` : "";

  for (const filePath of getBanFilePaths(serverDir, serverUuid)) {
    try {
      fs.mkdirSync(path.dirname(filePath), { recursive: true });
      fs.writeFileSync(filePath, payload, "utf-8");
    } catch (error: unknown) {
      const message = error instanceof Error ? error.message : String(error);
      return { success: false, message: `保存封禁列表失败 [${filePath}]: ${message}` };
    }
  }

  return { success: true, message: "封禁列表已保存" };
}

function appendParsedBans(content: string, target: LocalBanEntry[], seen: Set<string>): void {
  if (!content) {
    return;
  }

  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line) {
      continue;
    }
    const parts = line.split(/\s+/);
    if (parts.length === 0 || !parts[0]) {
      continue;
    }
    const guid = parts[0];
    if (seen.has(guid.toLowerCase())) {
      continue;
    }
    seen.add(guid.toLowerCase());

    let expiry = parts.length > 1 ? parts[1] : "";
    if (expiry === "-1") {
      expiry = "永久封禁";
    }
    const reason = parts.length > 2 ? parts.slice(2).join(" ") : "";
    target.push({ guid, time: expiry, reason });
  }
}
